storybookjs/storybook
# Angular: Fix Angular template error for props with a circular reference

Closes #25275, #16855





## What I did


For Angular stories, two templates are generated. One for the template that will be used for rendering and another that will be used for displaying source. The problem is the way objects are handled for the source template. This provides a fix for the circular reference case.

The `computesTemplateSourceFromComponent` attempts to replace input props with a value, instead of the variable. It relies on `createAngularInputProperty`, which uses `JSON.stringify` then applies some formatting replacements. The problem is that JSON.stringify does not handle circular references.

The solution I went with as an initial attempt is to use a replacer that replaces the circular reference with `"[Circular]"`. If there is a better convention then that can be changed. I wasn't sure what the best solution would be, so I just picked something.

As an example of my current solution, assume I have the following:
```ts
const MyStory = {
  render: args => {
    const obj = { a: 'b' };
    obj.ref = obj;
    return {
      props: {
        input1: obj,
      },
    };
  },
};
```

The resulting source template should be similar to the following:
```html
<story-comp [input1]="{a: "b", ref: "[Circular]"}" ></story-comp>
```

I added a test for this scenario `angular source decorator > with argTypes (from compodoc) > should handle circular object as stringified`. Also, I added tests for the `computesTemplateFromComponent` and updated a few test descriptions to not start with an uppercase, for consistency.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_



This is a matter of stringifying an object, so I don't think a manual testing process is necessary. I can try to find a spot to put a story, if needed. Running the unit test should be enough, but if you want to use a sandbox then you can pass an object to and input prop. The docs addon would also need to be loaded, so `sourceDecorator` gets called.

### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
